hudi-common/src/main/java/org/apache/hudi/common/data/HoodieListPairData.java

```diff
@@ -0,0 +1,175 @@
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collector;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static org.apache.hudi.common.function.FunctionWrapper.throwingMapToPairWrapper;
+import static org.apache.hudi.common.function.FunctionWrapper.throwingMapWrapper;
+
+/**
+ * In-memory implementation of {@link HoodiePairData} holding internally a {@link Stream} of {@link Pair}s.
+ *
+ * NOTE: This is an in-memory counterpart for {@code HoodieJavaPairRDD}, and it strives to provide
+ *       similar semantic as RDD container -- all intermediate (non-terminal, not de-referencing
+ *       the stream like "collect", "groupBy", etc) operations are executed *lazily*.
+ *       This allows to make sure that compute/memory churn is minimal since only necessary
+ *       computations will ultimately be performed.
+ *
+ * @param <K> type of the key in the pair
+ * @param <V> type of the value in the pair
+ */
+public class HoodieListPairData<K, V> extends HoodiePairData<K, V> {
+
+  private final Stream<Pair<K, V>> dataStream;
+
+  public HoodieListPairData(List<Pair<K, V>> data) {
+    this.dataStream = data.stream().parallel();
+  }
+
+  HoodieListPairData(Stream<Pair<K, V>> dataStream) {
+    this.dataStream = dataStream;
+  }
+
+  @Override
+  public List<Pair<K, V>> get() {
+    return dataStream.collect(Collectors.toList());
+  }
+
+  @Override
+  public void persist(String cacheConfig) {
+    // no-op
+  }
+
+  @Override
+  public void unpersist() {
+    // no-op
+  }
+
+  @Override
+  public HoodieData<K> keys() {
+    return new HoodieListData<>(dataStream.map(Pair::getKey));
+  }
+
+  @Override
+  public HoodieData<V> values() {
+    return new HoodieListData<>(dataStream.map(Pair::getValue));
+  }
+
+  @Override
+  public long count() {
+    return dataStream.count();
+  }
+
+  @Override
+  public Map<K, Long> countByKey() {
+    return dataStream.collect(Collectors.groupingBy(Pair::getKey, Collectors.counting()));
+  }
+
+  @Override
+  public HoodiePairData<K, Iterable<V>> groupByKey() {
+    Collector<Pair<K, V>, ?, List<V>> mappingCollector = Collectors.mapping(Pair::getValue, Collectors.toList());
+    Collector<Pair<K, V>, ?, Map<K, List<V>>> groupingCollector =
+        Collectors.groupingBy(Pair::getKey, mappingCollector);
+
+    Map<K, List<V>> groupedByKey = dataStream.collect(groupingCollector);
+    return new HoodieListPairData<>(groupedByKey.entrySet().stream()
+        .map(e -> Pair.of(e.getKey(), e.getValue())));
+  }
+
+  @Override
+  public HoodiePairData<K, V> reduceByKey(SerializableBiFunction<V, V, V> combiner, int parallelism) {
+    Map<K, java.util.Optional<V>> reducedMap = dataStream.collect(
+        Collectors.groupingBy(
+            Pair::getKey,
+            HashMap::new,
+            Collectors.mapping(Pair::getValue, Collectors.reducing(combiner::apply))));
```

Comment:
like this style over declaring the local vars in groupByKey() (var types are verbose and barely telling anything). better to align on the style.

Reply:
Explicit vars are necessary in some places since 1.8 javac can't infer types properly